draft-schinazi-quic-version-negotiation.md

```diff
@@ -112,42 +111,148 @@ versions as long as they are reasonably popular.  This is not generally an issue
+  in the subsequent Compatible Version fields, ordered by descending preference.
+  Note that the version in the Attempted Version field MUST be included in the
+  Compatible Version list to allow the client to communicate the attempted
+  version's preference.
+
+~~~
+ 0                   1                   2                   3
+ 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|                    Negotiated Version (32)                    |
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|                 Supported Version Count (i)                 ...
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|                    Supported Version 1 (32)                   |
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|                   [Supported Version 2 (32)]                  |
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+                               ...
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|                   [Supported Version N (32)]                  |
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+~~~
+{: #fig-server-tp title="Server Transport Parameter Format"}
+
+The content of each field is described below:
+
+Negotiated Version:
+
+: The version that the server chose to use for the connection. This field SHALL
+  be equal to the value of the Version field in the long header that carries
+  this transport parameter.
+
+Supported Version Count:
+
+: A variable-length integer specifying the number of Supported Version fields
+  following it. The server encodes all versions it supports in the subsequent
+  list, ordered by descending preference.
+
+
+Clients MAY include versions following the pattern `0x?a?a?a?a` in their
+Compatible Version list, and the server in their Supported Version list.
+Those versions are reserved to exercise version negotiation (see the Versions
+section of {{QUIC}}), and MUST be ignored when parsing these fields. On the
+other hand, the Received Negotiation Version list MUST be identical to the
+received Version Negotiation packet, so clients MUST NOT add or remove reserved
+version from that list.
+
+
+# Version Downgrade Prevention
+
+Clients MUST ignore any received Version Negotiation packets that contain the
+version that they initially attempted.
+
+When a server parses the client's `version_negotiation` transport parameter, if
+the `Received Negotiation Version Count` is not zero, the server MUST validate
+that it could have sent the Version Negotation packet described by the client.
+In particular, the server MUST ensure that there are no versions that it
+supports that are absent from the Received Negotiation Versions, and that the
+ordering matches the server's preference. This mitigates an attacker's ability
+to forge Version Negotiation packets to force a version downgrade.
```

Comment:
Suppose the client has preferences 3 > 2 > 1. However because
it thinks 3 is rare, it tends to first try 2. The attacker
supports 3 > 2 > 1.

In the ordinary case, the client would offer 2 and the server
would accept it.

However, consider what happens if the attacker forges a VN:

~~~
C->S: Ver = 2, Attempted Version = 2, VN = []
A->C: VN = 3, 2, 1
C->S: Ver = 3, Attempted Version = 3, VN = [3, 2, 1]
~~~
This handshake will succeed, and the attacker has succeeded
in *upgrading* you. I know that this doesn't seem like a real
attack, but formally they have influenced the negotiation.

Reply:
This wouldn't work because the client would drop a VN packet containing the version it tried. But I've added the previously attempted version field to tighten this even more.